service/storage_service.cc

```diff
@@ -6037,6 +6044,67 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
     }
 }
 
+future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+    // The coordinator does not execute global token metadata barrier before jumping to "cleanup" stage, so we need
+    // a barrier here to see the token metadata which is at least as recent as that of the sender.
+    auto& raft_server = _group0->group0_server();
+    co_await raft_server.read_barrier(&_abort_source);
+
+    shard_id shard;
+
+    { // do not keep token metadata longer than necessary in order not to block barriers
+        auto tm = _shared_token_metadata.get();
+        auto& tmap = tm->tablets().get_tablet_map(tablet.table);
+        auto *trinfo = tmap.get_tablet_transition_info(tablet.tablet);
+
+        // Check if the request is still valid.
+        // If there is mismatch, it means this cleanup was canceled and the coordinator moved on.
+        if (!trinfo) {
+            throw std::runtime_error(format("No transition info for tablet {}", tablet));
```

Comment:
Is this an error? It will just confuse operators.

Can't we just return?

Reply:
Q about leadership and rpc. Say a group0 leader lost leadership, but doesn't know it. What prevents it from continuing to send rpc based on outdated state? The rpc recipients will not know it's not a leader any more.

I think every rpc should be fenced.

Reply:
https://github.com/scylladb/scylladb/blob/cf37ab96f4cf5e429faaa2fc256bb295b1924b3a/service/storage_service.cc#L5699C12-L5699C14

Reply:
But then this call does not go through the generic topology RPC code. Why?

Reply:
Is relying on the term sufficient? It checks for no-loss-of-leadership, but it doesn't check for outdated messages. Outdated messages can happen if we send messages with timeout, the timeout expires, then we process them.

Reply:
> Is relying on the term sufficient? 

To detect message from the outdated leader, yet is is sufficient. 

> Outdated messages can happen if we send messages with timeout, the timeout expires, then we process them.

There is nothing wrong with that. It may always happen anyway. We send message with a timeout, it is processed and then timeout happen. Messages are idempotent.  They should not be re-ordered, but there is  a check for that as well.

Reply:
> Is this an error? It will just confuse operators.

It depends. If the leader is still valid, it is an unexpected condition and an error which will block topology change progress. If the reason for discrepancy is a race on lost leadership, it is benign. We don't know here which one is it.

Reply:
> Q about leadership and rpc. Say a group0 leader lost leadership, but doesn't know it. What prevents it from continuing to send rpc based on outdated state? The rpc recipients will not know it's not a leader any more.

I thought that raft protocol ensures that the old leader is notified one it connects to the quorum. If such a leader is able to contact a replica which knows he's not a leader, wouldn't raft protocol notify that old leader via reject message?

Reply:
It would. But it does not mean that there cannot be a time period in which an old leader thinks that it is still a leader, but the partitioned out majority already elected another leader. It is not easy to create such situation, but possible. Now the partitioning may heal and an old leader may send an RPC before it learns it is not a leader any longer.